### ISSUE TITLE
ngOnInit not firing when clicking browser's back button.

### ISSUE THREAD

**Issue by SU27Flanker** [angular/angular#26829]
## I'm submitting a...
<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Performance issue
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
[ ] Other... Please describe:
</code></pre>

## Current behavior
I'm having problems with routes that involve parameters. When I try to go back with the browser's back button from a route with a parameter like `person/:id` the ngOnInit of the previous component doesn't fire. This only happens when deploying the app to production.

## Expected behavior
The ngOnInit of the previous component should be called and render the component correctly.

## Minimal reproduction of the problem with instructions
Steps to reproduce the error with the angular's tutorial:

- Download the tutorial from [Angular's tutorial](https://angular.io/tutorial ) and run npm install
- Build the app with the command `ng build --prod --build-optimizer` or `ng build --prod`
- Upload the static files generated by the previous step to an nginx or apache server with the respective configuration file having the recomended lines of code in angular docs. [Angular Deployment](https://angular.io/guide/deployment)
- Click one of the top heros in the dashboard.
- Click browser's back button.

By following those steps you will see the top heros are not loaded in the dashboard which is exactly the same problem I'm having with my app. This is be cause ngOnInit is not called when going back specifically with the back button.  

## Environment

<pre><code>
Angular version: 6.0.0 and 7.0.1

Browser:
- [x] Chrome (desktop) version 69.0
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [x] Firefox version 63.0
- [x] Safari (desktop) version 12
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
 
For Tooling issues:
- Node version: 10.5.0
- Platform:  Mac, Linux, Windows

Others:
This problem doesn't accurs while using development mode ng serve and trying the app locally.
It only happens when deploying to a subdomain like "subdomain.domain.com" but works perfectly in a normal domain "domain.com". I'm not sure if this could be considered a normal behavior.

**Comment by jasonaden** [comment]
Thanks for the issue here, but it sounds more like a support request that should be sent over to StackOverflow. If you feel there's a specific error, please make a reproduction that can be quickly and easily run.

That being said, it sounds like you might just be re-using components when hitting *back*.

**Comment by SU27Flanker** [comment]
I don't know if there is a way to make this easier to reproduce, but I think it should work the same way when deploying to a subdomain than a domain. As I said before the problem seems to only happen with subdomains and just the browsers I specified in the issue. Some people will need to deploy to subdomains for some reason and there they gonna find this problem. I think it should be reviewed and fixed. 
P.D.: I posted this in stackoverflow before posting it here. But there was no answers for this.

**Comment by SU27Flanker** [comment]
> 
> 
> Thanks for the issue here, but it sounds more like a support request that should be sent over to StackOverflow. If you feel there's a specific error, please make a reproduction that can be quickly and easily run.
> 
> That being said, it sounds like you might just be re-using components when hitting _back_.

I don't know if there is a way to make this easier to reproduce, but I think it should work the same way when deploying to a subdomain than a domain. As I said before the problem seems to only happen with subdomains and just the browsers I specified in the issue. Some people will need to deploy to subdomains for some reason and there they gonna find this problem. I think it should be reviewed and fixed.
P.D.: I posted this in stackoverflow before posting it here. But there was no answers for this.

**Comment by glemiere** [comment]
Actually it’s not just ngOnInit. The whole lifecycle is broken on production build when deploying to a subdomain.

I’ll dig into Angular code and see if I can figure something out.

**Comment by trotyl** [comment]
> As I said before the problem seems to only happen with subdomains and just the browsers I specified in the issue.

Both StackBlitz and GitHub pages are subdomain, or where else are you going to repro it?

**Comment by SU27Flanker** [comment]
> 
> 
> > As I said before the problem seems to only happen with subdomains and just the browsers I specified in the issue.
> 
> Both StackBlitz and GitHub pages are subdomain, or where else are you going to repro it?

To get the problem I made a digitalocean's droplet with ubuntu server, then installed nginx and made a configuration file for the virtual host with the line specified in the docs `try_files $uri $uri/ /index.html;`. I also did the same but with an apache server instead of nginx and used the respective .htaccess file. In both cases I had the problem I described in the main post.

**Comment by glemiere** [comment]
@jasonaden please re-open. This issue has been plaguing Angular apps in production for the past two years now.

**Comment by ratheeps** [comment]
I'm also facing the same issue!

**Comment by shailesh-sangekar** [comment]
I am facing same issue with angular 7 i.e. ngOnInit event is not firing when clicking browser's back for forward button. @jasonaden  Can you please provide solution or work around for this?

**Comment by Kayes-Islam** [comment]
Facing same issue. Please provide solution.

**Comment by Venkatesanchuchu** [comment]
Facing Same issue . While calling same component with different params

**Comment by glemiere** [comment]
@jasonaden please re-open.

**Comment by gustavshf** [comment]
Facing the same issue, can't find a workaround.

**Comment by gustavshf** [comment]
The only workaround that worked. Snippet taken from #7722 and modified.

```
constructor(private readonly _router: Router,
        _applicationRef: ApplicationRef,
        zone: NgZone) {
            _router.events.subscribe((uri)=> {
                zone.run(() => _applicationRef.tick());
            });
    }
```

**Comment by zelishe** [comment]
Here is a new version of workaround for Angular 8. setTimeout is used to avoid "ApplicationRef.tick is called recursively" error. There is no check for Safari browser, because this issue happens in other browsers as well.

```
  constructor(
    private router: Router,
    private applicationRef: ApplicationRef,
    private zone: NgZone
  ) {
    router.events.subscribe(() => {
      zone.run(() =>
        setTimeout(() => {
          this.applicationRef.tick();
        }, 0)
      );
    });
  }
```

**Comment by l0co** [comment]
The same issue with `7.3.9` + cordova `9.0.0` using following commands to run the app in the device:

```
ng build --base-href "./"
cordova run --debug android
```

**Comment by aman395jain** [comment]
This issue is still exist.

**Comment by codotronix** [comment]
This issue still exists. Anyone found any work around please ?

**Comment by l0co** [comment]
Looking on the history, in my opinion they stopped reading this ticket after marking it as closed :/

**Comment by glemiere** [comment]
This is so frustrating.

This is a huge issue and I was hoping a framework maintained by GOOGLE would be assigned enough resources to properly test, listen to the community, and fix anything CRITICAL a little faster than that.

I am very disappointed in Angular right now.

**Comment by thomaswmanion** [comment]
The best workaround I've found (if you're globally OK with these settings) is to do this on application start:

``` 
    this.router.onSameUrlNavigation = 'reload';
    this.router.routeReuseStrategy.shouldReuseRoute = () => false;
```

**Comment by sarunint** [comment]
I've read through this issue and I have not seen any reproduction of the issue.

Guys, if you need the issue to be looked at, at least provide a StackBlitz, or a GitHub repository.

**Comment by glemiere** [comment]
@sarunint a StackBlitz or a repo wouldn't be helpful. You just need to generate a new project and host it on a subdomain.

**Comment by sarunint** [comment]
I'll repeat what @trotyl said:

> Both StackBlitz and GitHub pages are subdomain, or where else are you going to repro it?

If you can't reproduce it StackBlitz, then it might be your problem with deployment. In that case, provide a `Dockerfile` with the source code, then.

**Comment by codotronix** [comment]
The issue is present only in real devices, not in browser emulator, but
when packed the entire angular app with Apache Cordova and run the .apk on
mobile. This issue is real.

**Comment by Helen2011** [comment]
Facing same issue. Please provide solution.

**Comment by Danieliverant** [comment]
any news on that? any working workaround?

**Comment by Danieliverant** [comment]
My use case and my solution:

**Use case:**
Two modules: Auth and Dashboard.
The user perform login in AuthModule get UserToken and redirected to the DashboardModule.
The DashboardModule has a header, sidebar, and main. 
The main has a guard that get AccessToken and redirect the user to an inner page (dashboard/home, dashboard/news, etc...).

problem: when the user clicks "back" or "slide back" (on iPhone) the guard that need to get the AccessToken and redirect to an inner page not activated and the user see a blank page with a header and sidebar but without content (main).

**Solution** - inside the HeaderComponent OnInit cycle:
`this.router.events
      .pipe(
        takeUntil(this.unsubscriber$)
      )
      .subscribe(e => {
        if (e instanceof NavigationEnd && e.url === PageURLs.DASHBOARD) {
          this.location.back();
        }
      });`

where PageURLs.DASHBOARD = the URL of the problematic page, in my case: "/dashboard".

**Comment by Helen2011** [comment]
I just upgrade to angular 8.2.8, looks like it's fixed on 8.2.8.

**Comment by Rubenazo** [comment]
> I just upgrade to angular 8.2.8, looks like it's fixed on 8.2.8.

Just updated my project to Angular 8.2.8 for this, the problem is still there

**Comment by elhanafih** [comment]
Same issues is still there, i got it when i click on the browser's back button, i got redirected to a route as i wanted but the html page shown is incorrect, did not belongs to the desired route, its visible briefly and disapear, and it did not fire its ngOnInit method, this is happen only on PROD and Node server.
